source/common/common/inline_map_registry.h

```diff
@@ -0,0 +1,389 @@
+    void insert(InlineKey handle, TPtr value) {
+      if (handle.registryId() != registry_.registryId()) {
+        return;
+      }
+
+      ASSERT(handle.inlineId() < registry_.registrationNum());
+      resetInlineMapEntry(handle.inlineId(), std::move(value));
+    }
+
+    // Remove the entry for the given key. If the key is not found, do nothing.
+    void remove(absl::string_view key) {
+      // Compute the hash value for the key and avoid computing it again in the lookup.
+      const size_t hash_value = absl::Hash<absl::string_view>()(key);
+
+      if (auto entry_id = inlineLookup(key, hash_value); entry_id.has_value()) {
+        resetInlineMapEntry(*entry_id);
+      } else {
+        normal_entries_.erase(key);
+      }
+    }
+
+    // Remove the entry for the given handle. If the handle is not valid, do nothing.
+    void remove(InlineKey handle) {
+      if (handle.registryId() != registry_.registryId()) {
+        return;
+      }
+      resetInlineMapEntry(handle.inlineId());
+    }
+
+    void iterate(std::function<bool(absl::string_view, T*)> callback) const {
+      for (const auto& entry : normal_entries_) {
+        if (!callback(entry.first, entry.second.get())) {
+          return;
+        }
+      }
+
+      for (const auto& id : registry_.registrationMap()) {
+        ASSERT(id.second < registry_.registrationNum());
+
+        auto entry = inline_entries_[id.second];
+        if (entry == nullptr) {
+          continue;
+        }
+
+        if (!callback(id.first, entry)) {
+          return;
+        }
+      }
+    }
+
+    uint64_t size() const { return normal_entries_.size() + inline_entries_size_; }
+
+    bool empty() const { return size() == 0; }
+
+    // Get the registry id that the inline map created by.
+    uint64_t registryId() const { return registry_.registryId(); }
+
+    ~InlineMap() {
+      for (uint64_t i = 0; i < registry_.registrationNum(); ++i) {
+        auto entry = inline_entries_[i];
+        if (entry != nullptr) {
+          delete entry;
+        }
+      }
+      memset(inline_entries_, 0, registry_.registrationNum() * sizeof(TPtr));
+    }
+
+  private:
+    friend class InlineMapRegistry;
+
+    static std::unique_ptr<InlineMap> create(InlineMapRegistry& registry) {
+      return std::unique_ptr<InlineMap<T>>(new (
+          (registry.registrationMap().size() * sizeof(InlineMap<T>::TPtr))) InlineMap(registry));
+    }
+
+    InlineMap(InlineMapRegistry& registry) : registry_(registry) {
+      // Initialize the inline entries to nullptr.
+      memset(inline_entries_, 0, registry_.registrationNum() * sizeof(TPtr));
+    }
+
+    T* lookupImpl(absl::string_view key) const {
+      // Compute the hash value for the key and avoid computing it again in the lookup.
+      const size_t hash_value = InlineMapRegistry::Hash()(key);
+
+      // Because the normal string view key is used here, try the normal map entry first.
```

Comment:
again I think 'normal' is a little confusing. Suggest 'inline' and 'dynamic' to distinguish the cases.

Shouldn't we be checking inline keys first?

Reply:
The order doesn't matter when doing the searching. Because the key will only be store in on location. Because normal/dynamic key is used, so the normal/dynamic entries will be tried first.
If it's found, then we can reduce unnecesary searching.